compiler/utils/var.mli

```diff
@@ -0,0 +1,73 @@
+(* This file is part of the Catala compiler, a specification language for tax
+   and social benefits computation rules. Copyright (C) 2020-2022 Inria,
+   contributor: Louis Gesbert <louis.gesbert@inria.fr>
+
+   Licensed under the Apache License, Version 2.0 (the "License"); you may not
+   use this file except in compliance with the License. You may obtain a copy of
+   the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+   WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+   License for the specific language governing permissions and limitations under
+   the License. *)
+
+open Astgen
+
+(** {1 Variables and their collections} *)
+
+(** This module provides types and helpers for Bindlib variables on the
+    [Astgen.gexpr] type *)
+
+type 'e expr = 'e
+  constraint 'e = ([< desugared | scopelang | dcalc | lcalc ], 't) gexpr
+(** Subtype of Astgen.gexpr where variables are handled *)
+
+type 'e var = 'e expr Bindlib.var
+type 'e t = 'e var
+type 'e vars = 'e expr Bindlib.mvar
+
+val make : string -> 'e t
+val compare : 'e t -> 'e t -> int
+val eq : 'e t -> 'e t -> bool
+
+val translate : 'e1 t -> 'e2 t
+(** Needed when converting from one AST type to another. See the note of caution
+    on [Bindlib.copy_var]. *)
+
+(** Wrapper over [Set.S] but with a type variable for the AST type parameters.
+    Extend as needed *)
+module Set : sig
```

Comment:
Couldn't that be expressed with `Set.S with type elt = ...` ?

Reply:
The issue here is that `Set.elt` doesn't allow a type parameter.So we would have to go through a functor instanciation for each of the `'a` in `'a expr`.

All these shenanigans are just to instanciate a `Set.S` with `type 'a t` and matching `type 'a elt`. Unfortunately that forces us to rewrite the whole interface and use some tricks in the implementation…

An alternative, given we want to be able to write polymorphic functions that will handle `'a variables` and use sets or maps internally, would be to use first-class modules and instanciate them each time. But I think that would lead to much harder to read interfaces.

For example, in `dcalc/ast.ml`, we have
```ocaml
val free_vars_expr : 'm marked_expr -> 'm expr Var.Set.t
```
We could make this function return a first-class module instead, buuut:
```ocaml
module type SetValue = sig
  include Set.S
  val ret: elt
end
val free_vars_expr : 'm marked_expr -> (module SetValue with type elt = 'm expr)
```

Starts to feel like OOP at this point…

Reply:
Aaaah OK thanks for the explanation. Yeah let's things as they are now.